File: 125_algorithms/_exercises/templates/_algorithms_challenges/leetcode/leetCode/String/05_LongestPalindromicSubstring.cpp
/*
 * @Author: xuezaigds@gmail.com
 * @Last Modified time: 2016-04-04 21:41:17
 */

c.. Solution {
public:
    string longestPalindrome(string s) {
        __(s__"")   r_ "";
        else __(s.size() __ 1) r_ s;
        int s_len = s.size();
        int max_begin=0, max_end=0;
        int pos = 0;
        _____(pos < s_len){
            // No need to check the remainming, pruning here
            __(max_end-max_begin >= (s_len-pos) * 2 - 1){
                break;
            }
            int left = pos, right = pos+1;
            _____(right < s_len && s[right] __ s[right-1]){
                right ++;
            }
            pos = right;
            _____(left-1>=0 && right<s_len && s[left-1]__s[right]){
                left -= 1;
                right += 1;
            }
            __(right - left > max_end - max_begin){
                max_begin = left;
                max_end = right;
            }
        }
        r_ s.substr(max_begin, max_end-max_begin);
    }
};

/*
""
"a"
"aa"
"dcc"
"aaaa"
"cccd"
"ccccdc"
"abcdefead"
*/

File: 125_algorithms/_exercises/templates/_algorithms_challenges/leetcode/leetCode/String/05_LongestPalindromicSubstring.py
#! /usr/bin/env python
# -*- coding: utf-8 -*-
# @Author: xuezaigds@gmail.com
# @Last Modified time: 2016-04-04 17:39:46


c.. Solution o..
    # Easy to understand.  Refer to
    # https://leetcode.com/discuss/32204/simple-c-solution-8ms-13-lines
    ___ longestPalindrome  s
        __ n.. s:
            r_ ""
        s_len = l..(s)
        __ s_len __ 1:
            r_ s
        max_start, max_end = 0, 1   # Make sure s[start:end] is palindrome
        i = 0
        _____ i < s_len:
            # No need to check the remainming, pruning here
            __ max_end - max_start >= (s_len-i) * 2 - 1:
                break
            left, right = i, i+1
            # Skip duplicate characters i.
            _____ right < s_len and s[right-1] __ s[right]:
                right += 1
            i = right
            _____ left-1 >= 0 and right < s_len and s[left-1] __ s[right]:
                left -= 1
                right += 1
            __ right-left > max_end-max_start:
                max_start = left
                max_end = right
        r_ s[max_start:max_end]

"""
""
"a"
"aa"
"dcc"
"aaaa"
"cccd"
"ccccdc"
"abcdefead"
"""
